### ISSUE TITLE
Inconsistent “dtick” when using ‘auto’ tickmode

### ISSUE THREAD

**Issue by OmriAroesti** [plotly/plotly.js#5541]
I am using plotly js together with React (react-plotly.js), and try to overcome inconsistent behavior that causes a display bug.

I am using a monthly time series with a range slider. All dates in the chart are at a monthly interval.
When at a certain threshold of zoom level all is displayed fine - with tick interval of 1 month, inline with the data.
However, when zooming in past that threshold, suddenly the interval switches to 28 days, and causes misalignment and the wrong date being displayed.

Normal display
![d](https://user-images.githubusercontent.com/67684817/110619728-d13b3180-81a0-11eb-91ea-3f49f690ffc5.png)

Wrong display
![ww](https://user-images.githubusercontent.com/67684817/110619750-d7311280-81a0-11eb-8de7-bc304c14dc38.png)

As you can see, in this case, the 1st of January appears as Dec 31 (28 days gap).
I could set dtick to ‘1M’ to force the intervals to stay at 1 month apart, but that would cause all labels to appear when zoomed out, cluttering the view, so I’d like to keep the auto behavior.

**Comment by OmriAroesti** [comment]
Is there any way to handle this? Seems like a lot of people would encounter that type of issue

**Comment by alexcjohnson** [comment]
@OmriAroesti I'm not able to reproduce the 28-day `dtick` via automated logic on zoom, in my tests it goes from 1 month directly to 14 days. Can you reproduce this effect in a codepen so we can see it?

That said there's a potentially interesting feature here, to set a minimum `dtick`, either explicitly or automatically based on the data. When the data traces all have 1-month resolution there's really no benefit to finer ticks, and it's particularly confusing for grouped bars where horizontal position within the group doesn't encode anything about the time.

**Comment by OmriAroesti** [comment]
@alexcjohnson 
[Link to codesandbox](https://codesandbox.io/s/react-plotly-example-forked-jkund?file=/src/Plot.js)
1.Use the range slider to get January '12 into view
![Screen Shot 2021-03-15 at 21 08 04](https://user-images.githubusercontent.com/67684817/111207646-8e2ef300-85d2-11eb-8c19-3cc4d6a94fcc.png)
2. Zoom in until instead of Jan 1 and Feb 1, you get Jan 3 and Jan 31.
![Screen Shot 2021-03-15 at 21 10 44](https://user-images.githubusercontent.com/67684817/111208031-ec5bd600-85d2-11eb-9ca2-ecb9132d2496.png)

I only add the month days for this demonstration, but for my use case I only want to display the month and year (as this is the gaps between items). In those cases, January would repeat (instead of correctly displaying as February).

I suspect the difference between your threshold being 14 days is that in my case there are multiple bars.

**Comment by alexcjohnson** [comment]
Ah, got it - thanks. Here's one where I managed to get a 29-day `dtick` by including Feb and Mar of a leap year only. https://codepen.io/alexcjohnson/pen/MWbRKEK?editors=0010

We appear to have some logic specifically for grouped bars, that we don't allow an automatic dtick less than the minimum bar spacing. We're going to have to figure out how to distinguish a real 28- or 29-day minimum spacing from a one-month minimum spacing. Same in principle may apply to n-month spacing (including yearly), though I bet it's going to be hard to find cases where this is obviously problematic. I guess it would cover 99% of cases if we simply took any such result that's at least 28 days and rounded up to the nearest number of months?
